### ISSUE TITLE
Job Kill and Job Timeout issues

### ISSUE THREAD

**Issue by biguphpc** [rundeck/rundeck#1238]
Killing a job doesn't kill the whole process tree, this is an issue because most jobs are run through **/bin/sh** and thus only **sh** is killed, not the actual process.

The job timeout has the same issues. When the job timeouts the actual process keeps running.

To reproduce:
1. Create a job with a command step
2. Set the command as (for example)  `python -c """f = open('/dev/random', 'r'); print f; f.read()"""`
3. Set a 1m timeout

Job runs: `ps -ef`
`rundeck  59337  1219  /bin/sh -c python -c "f = open('/dev/random', 'r'); print f; f.read()"
rundeck  59340  59337 python -c f = open('/dev/random', 'r'); print f; f.read()`

After a manual kill or a timeout:
`rundeck  59340  python -c f = open('/dev/random', 'r'); print f; f.read()`

Notice only /bin/sh has been killed but the other process is still running.

We also experienced issues where a timeout will not be 'detected' and the job keeps running for multiple days until we manually killed it inside Rundeck.

**Comment by juziel** [comment]
I am experiencing the same issue. A pkill -f before the script helps in the case of retries.

**Comment by sea-lmarchal** [comment]
I'm having trouble finding in the rundeck code where the timeout/kill is handled. If somebody can give me pointer on the classes i'm looking for I'd be happy to fix it and create a pull request.

**Comment by sylvainr** [comment]
This is indeed a serious problem.

The consequence is that a scheduled job that freezes (and does not timeout) it may prevent other jobs to actually happen going forward (when the job option is that it should not allow multiple to run at the same time) until someone "kills" the job manually.

I've tried to find the place in the code where time timeout is supposedly used, but could not find much. Any pointers would be appreciated.

**Comment by mathieuchateau** [comment]
+1, also this issue if real command is started with sudo within job step

**Comment by biguphpc** [comment]
I'm also having trouble finding where the timeout is handled in the code. I'll be happy to work on a fix if somebody could provide some pointers.

**Comment by cmbrehm** [comment]
I believe it's the stopAgent() method in 
```core/src/main/java/com/dtolabs/rundeck/core/utils/SSHAgentProcess.java``` that you're looking for

**Comment by gschueler** [comment]
Hi everyone:

first point: the SSHAgentProcess.java is just for the "ssh-agent" feature of the SSH plugin.

For the issue of killing a process tree, there needs to be some enhancement to the way that the node executor can keep track of the process ID and later kill it and any related processes if the job is cancelled.  We could do this by adding something to the ExecutionContext where a step can register a "cleanup" task in case it is aborted.

Right now, "Kill job" and "timeout" cause the same effect: the main thread for the execution is interrupted, which *may* allow some of the steps in the workflow to notice that the job was cancelled and handle cleanup.  I say *may* because if you are using multiple threads to execute across nodes, those threads won't necessarily be interrupted, but new threads will not be started.

Finally, when executing a step remotely on another node via SSH, I'm not sure how best to handle a "kill".  We'd need acquire the PID on the remote node.

To provide some pointers:

* "Kill" will tell Quartz to interrupt the quartz job, which calls "interrupt()" on ExecutionJob.groovy
* "Timeout" delay is tracked by ExecutionJob, which then calls interrupt() on itself if the delay is elapsed.

Once interrupt() is called, `abort()` is called on the WorkflowExecutionServiceThread.java class, which actually does the `Thread.interrupt()` on itself.

At that point the core workflow executor strategy will be able to notice if the thread was interrupted.  In the single-threaded case this may be a SequentialNodeDispatcher, which will check interrupted status as it loops through nodes.  The ParallelNodeDispatcher is currently a wrapper for the Ant ParallelTask.  Also, parallel workflow strategy or the Orchestrator strategy behave differently.

I like the idea of having a node executor (SSH or local exec, or any kind of plugin) be able to register a callback to execute a kill command if necessary.

**Comment by biguphpc** [comment]
Thanks for the pointers, it's appreciated. I looked around the code and from what I see, when a job is ran locally (no ssh) here are the steps:

1. ExecutionJob.groovy calls executionService.executeAsyncBegin()
2. ExecutionService.executeAsyncBegin() creates a WorkflowExecutionServiceThread and starts it
3. WorkflowExecutionServiceThread calls WorkflowExecutor.executeWorkflow()
4. the WorkflowExecutor ends up calling LocalExecNodeStepPlugin
5. LocalExecNodeStepPlugin uses ScriptExecUtil.
6. ScriptExecUtil executes runLocalCommand() that uses Runtime.exec()

Long story short, local jobs are never killed since Process.destroy() is never called.
Even if Process.destroy() was called, Java does not kill sub-processes that way. Since most likely 'sh' is the process running it would not kill the commands run after sh. The best would be to use JNI on Windows and on Unix plateforms call the 'kill' command to kill the whole process group.

The other issue is the timeout, I think there's threading issue in ExecutionJob.groovy that might not allow to detect timeouts in every situation.

**Comment by gschueler** [comment]
@biguphpc Yes that's mostly correct, but I'll point out that LocalExecNodeStepPlugin will only be used for the "Local Command" plugin step.  The built-in "command" steps will invoke ExecNodeStepExecutor (NodeStepExecutor is the built-in type equivalent to a NodeStepPlugin), and will end up going through the LocalNodeExecutor (locally) or JschNodeExecutor (for SSH).  LocalNodeExecutor wraps Ant's ExecTask.  In the end it similarly uses `Runtime.exec()`.

**Comment by new23d** [comment]
@gschueler, regarding your comment "Finally, when executing a step remotely on another node via SSH, I'm not sure how best to handle a "kill". We'd need acquire the PID on the remote node."

I don't think Rundeck should take the responsibility for that. The command set to run on the remote node may have chosen to go into background for one case. Withdrawal of the connection from Rundeck should suffice and must mimic the usual behaviour of consoles dying, i.e. sending the HUP signal. If HUP is trapped on the remote node, well and good; if not, Rundeck mustn't do anything special and over the top, and leave the HUP to cause natural reactions.

**Comment by yogeshkumark** [comment]
+5

**Comment by fdevans** [comment]
If you are still experiencing this issue on the current version of Rundeck please let us know by opening a new Issue request with version and steps to reproduce.  If you are an Enterprise customer our support personnel are ready to help.
Thank you,
The Rundeck Team
